### PR TITLE
Fix dynamically added/removed animated styles on iOS/Android

### LINKED ISSUE #5296: Re-rendering component with inline SharedValues causes style loss
### Description

When using inline `useSharedValue` animated components lose styling when the component rerenders due to a state update or fast-refresh.

https://github.com/software-mansion/react-native-reanimated/assets/3946701/7435d3c4-fc6e-4413-b78a-c1d369455c2c

### Steps to reproduce

- Clone repo.
  - It was created by `npx react-native@latest init` and following the Getting Started and Your First Animation guides.

- Click the button and observe
  - Purple rectangle flickers (it disappears for 1 frame)
  - Red text turns gray

- Reload app

- Modify a component and observe
  - Purple rectangle disappears
  - Red text turns gray

### Snack or a link to a repository

https://github.com/marklawlor/reanimated-inline

### Reanimated version

3.5.4

### React Native version

0.72.6

### Platforms

Android

### JavaScript runtime

None

### Workflow

React Native

### Architecture

Paper (Old Architecture)

### Build type

Debug mode

### Device

Android emulator

### Device model

Pixel XL API 33

### Acknowledgements

Yes

### PR DESCRIPTION
**Pull request by tjzel**
## Summary

Fixes #5296.

1. Since this is potentially a JS-heavy fix, I'll gladly listen to suggestions and receive help with how to profile it.

2. I personally dislike how those changes look like. I would much more prefer to simplify some logic present in `createAnimatedComponent` and its devil spawn, reducing some code, instead of building on top of it - although I'm not sure if we need it at the moment.

Just watch the video to understand what is happening:

<table>
<tr><td>BEFORE (buggy)</td><td>AFTER (desired)</td></tr>
<tr><td>

https://github.com/software-mansion/react-native-reanimated/assets/40713406/07495711-a909-4499-89b8-a0be17f3d8ef

</td><td>

https://github.com/software-mansion/react-native-reanimated/assets/40713406/47cbace1-bf98-4011-8b4c-79bc6f6a80d4

</td></tr>
</table>

## Test plan

💣 

## Note

JS-Reanimated (Web implementation) still has those issues and I will produce a separate solution for it in a finite amount of time.

### CODE REVIEW

**Comment by Latropos** [comment]
Wanna add "closing #5296"?

File: app/src/examples/DynamicStylesExample.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import React from 'react';
+
+import Animated, {
+  useAnimatedStyle,
+  useSharedValue,
+  withTiming,
+} from 'react-native-reanimated';
+
+export default function DynamicStylesExample() {
+  const [extraStyle, setState] = React.useState(false);
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:12]:
```suggestion
  const [extraStyle, setExtraStyle] = React.useState(false);
```

File: app/src/examples/DynamicStylesExample.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  const [rerender, setRerender] = React.useState(false);
+  const widthShared = useSharedValue(80);
+  const backgroundColorShared = useSharedValue<string>('rgb(255,127,63)');
+  const transformScaleYShared = useSharedValue(1);
+  const animatedStyle = useAnimatedStyle(() => ({ width: widthShared.value }));
+  const animatedStyle2 = useAnimatedStyle(() => {
+    return {
+      backgroundColor: backgroundColorShared.value,
+    };
+  });
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:22]:
```suggestion
  const animatedWidthStyle = useAnimatedStyle(() => ({ width: widthShared.value }));
  const animatedBackgroundColorStyle = useAnimatedStyle(() => {
    return {
      backgroundColor: backgroundColorShared.value,
    };
  });
```

File: app/src/examples/DynamicStylesExample.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+  function handleChangeProps() {
+    setState(!extraStyle);
+  }
+  function handleAnimate() {
+    widthShared.value = withTiming(Math.random() * 160);
+    backgroundColorShared.value = withTiming(
+      `rgb(${Math.random() * 256},${Math.random() * 256},${
+        Math.random() * 256
+      })`
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:32]:
```suggestion
      const r = Math.random() * 255;
      const g = Math.random() * 255;
      const b = Math.random() * 255;
      `rgb(${r},${g},${b})`
```

File: app/src/examples/DynamicStylesExample.tsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+          <Text>Same length array</Text>
+          <Animated.View
+            accessible={rerender}
+            style={[
+              styles.box,
+              { width: widthShared },
+              extraStyle
+                ? {
+                    backgroundColor: backgroundColorShared,
+                  }
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:57]:
```suggestion
                ? { backgroundColor: backgroundColorShared }
```

File: app/src/examples/DynamicStylesExample.tsx [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          <Text>Two animated styles independent of state</Text>
+          <Animated.View
+            accessible={rerender}
+            style={[styles.box, animatedStyle, animatedStyle2]}
+          />
+        </View>
+      </View>
+
+      <Text>
+        {extraStyle ? 'extra animated style' : 'no extra animated style'}
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:160]:
```suggestion
        {extraStyle ? '' : 'no '}extra animated style
```

File: app/src/examples/DynamicStylesExample.tsx [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  return (
+    <View style={styles.container}>
+      <Text>State: {String(rerender)}</Text>
+      <View style={styles.row}>
+        <View style={styles.container}>
+          <Text>InlineStyles</Text>
+
+          <Text>Same length array</Text>
+          <Animated.View
+            accessible={rerender}
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:50]: Why do we use `accessible` prop here? If you want to force re-render, usually we use `key` prop.

File: app/src/examples/DynamicStylesExample.tsx [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+import Animated, {
+  useAnimatedStyle,
+  useSharedValue,
+  withTiming,
+} from 'react-native-reanimated';
+
+export default function DynamicStylesExample() {
+  const [extraStyle, setState] = React.useState(false);
+  const [rerender, setRerender] = React.useState(false);
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:13]: Let's separate React state from shared state by adding a newline here.
```suggestion
  const [rerender, setRerender] = React.useState(false);

```

File: app/src/examples/DynamicStylesExample.tsx [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+  const widthShared = useSharedValue(80);
+  const backgroundColorShared = useSharedValue<string>('rgb(255,127,63)');
+  const transformScaleYShared = useSharedValue(1);
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:16]: Let's separate shared state from animated styles by adding a newline here.
```suggestion
  const transformScaleYShared = useSharedValue(1);

```

File: app/src/examples/DynamicStylesExample.tsx [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+  const animatedStyle = useAnimatedStyle(() => ({ width: widthShared.value }));
+  const animatedStyle2 = useAnimatedStyle(() => {
+    return {
+      backgroundColor: backgroundColorShared.value,
+    };
+  });
+
+  function handleChangeProps() {
+    setState(!extraStyle);
+  }
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:26]:
```suggestion
  }

```

File: app/src/examples/DynamicStylesExample.tsx [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+  const transformScaleYShared = useSharedValue(1);
+  const animatedStyle = useAnimatedStyle(() => ({ width: widthShared.value }));
+  const animatedStyle2 = useAnimatedStyle(() => {
+    return {
+      backgroundColor: backgroundColorShared.value,
+    };
+  });
+
+  function handleChangeProps() {
+    setState(!extraStyle);
```
Comment (tomekzaw) [inline app/src/examples/DynamicStylesExample.tsx:25]:
```suggestion
    setState(extraStyle => !extraStyle);
```

File: app/src/examples/DynamicStylesExample.tsx (hunk 6 above)
Comment (tjzel) [reply app/src/examples/DynamicStylesExample.tsx:50]: Key prop causes the remount of the component.

**Review by bartlomiejbloniarz — APPROVED** [review]
🧨

File: src/createAnimatedComponent/PropsFilter.tsx [hunk 11]
```diff
@@ -61,6 +72,7 @@ export class PropsFilter implements IPropsFilter {
           Object.keys(animatedProp.initial.value).forEach((initialValueKey) => {
             props[initialValueKey] =
               animatedProp.initial?.value[initialValueKey];
+            // TODO - refacotr, since `viewsRef` is only present on Web
```
[the remark this answers is not in the document]
Comment (Latropos) [reply src/createAnimatedComponent/PropsFilter.tsx:75]: I use VSCode with https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker and recommend it

[the remark this answers is not in the document]
Comment (tjzel) [reply src/createAnimatedComponent/PropsFilter.tsx:75]: I keep it 😈

**Pull request merged**